core/trino-main/src/main/java/io/trino/operator/join/LookupJoinPageBuilder.java

```diff
@@ -98,44 +100,50 @@ public void appendNullForBuild(JoinProbe probe)
+        // Unwrap them to reduce overhead of further processing.
+        return filteredProbeBlock.isLoaded() ? filteredProbeBlock.getLoadedBlock() : filteredProbeBlock;
+    }
+
     public Page build(JoinProbe probe)
     {
-        int[] probeIndices = probeIndexBuilder.toIntArray();
-        int length = probeIndices.length;
-        verify(buildPageBuilder.getPositionCount() == length);
+        int outputPositions = probeIndexBuilder.size();
+        verify(buildPageBuilder.getPositionCount() == outputPositions);
 
         int[] probeOutputChannels = probe.getOutputChannels();
         Block[] blocks = new Block[probeOutputChannels.length + buildOutputChannelCount];
-        for (int i = 0; i < probeOutputChannels.length; i++) {
-            Block probeBlock = probe.getPage().getBlock(probeOutputChannels[i]);
-            if (!isSequentialProbeIndices || length == 0) {
-                blocks[i] = probeBlock.getPositions(probeIndices, 0, probeIndices.length);
-            }
-            else if (length == probeBlock.getPositionCount()) {
-                // probeIndices are a simple covering of the block
-                verify(probeIndices[0] == 0);
-                verify(probeIndices[length - 1] == length - 1);
-                blocks[i] = probeBlock;
+        Page probePage = probe.getPage();
+        if (!isSequentialProbeIndices || outputPositions == 0) {
+            int[] probeIndices = probeIndexBuilder.toIntArray();
+            for (int i = 0; i < probeOutputChannels.length; i++) {
+                blocks[i] = unwrapLoadedBlocks(probePage.getBlock(probeOutputChannels[i]).getPositions(probeIndices, 0, probeIndices.length));
             }
-            else {
-                // probeIndices are sequential without holes
-                verify(probeIndices[length - 1] - probeIndices[0] == length - 1);
-                blocks[i] = probeBlock.getRegion(probeIndices[0], length);
-            }
-
-            // Lazy blocks (e.g used in filter condition) could be loaded during filter evaluation.
-            // Unwrap them to reduce overhead of further processing.
-            if (blocks[i].isLoaded()) {
-                blocks[i] = blocks[i].getLoadedBlock();
+        }
+        else {
+            // probeIndices are sequential without holes
+            int startRegion = probeIndexBuilder.getInt(0);
+            verify(previousPosition - startRegion == outputPositions - 1);
+            // probeIndices are a simple covering of the block, output the probe block directly
+            boolean outputProbeBlocksDirectly = startRegion == 0 && outputPositions == probePage.getPositionCount();
+
+            for (int i = 0; i < probeOutputChannels.length; i++) {
+                Block block = probePage.getBlock(probeOutputChannels[i]);
+                if (!outputProbeBlocksDirectly) {
+                    // only a subregion of the block should be output
+                    block = block.getRegion(startRegion, outputPositions);
```

Comment:
why not unwrap it when it's loaded?

Reply:
The unwrapping will occur after the region selection outside of the conditional here (which handles unwrapping both scenarios).